### PR TITLE
ADD RockFluidOrganizationInterpretation

### PR DESCRIPTION
**Pull request by untereiner**
(Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.)

What does this implement/fix? Explain your changes.
---------------------------------------------------
RockFluidOrganizationInterpretation

Does this close any currently open issues?
------------------------------------------
#107 

Any relevant logs, error output, etc?
-------------------------------------
(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

Any other comments?
-------------------
…

Where has this been tested?
---------------------------
**Operating System:**macos / gcc-8.3

**Platform:** …

### CODE REVIEW

**Comment by untereiner** [comment]
Hi Philippe,
I started the implementation of RockFluidOrganizationInterpretation.
For now I have no loaded grids into RockFluidOrganizationInterpretation objects and I do not know why

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+		/**
+		* Creates an instance of this class in a gsoap context.
+		* @param orgFeat			The feature the instance interprets. It must be a stratigraphic organization feature.
+		* @param guid				The guid to set to the interpretation. If empty then a new guid will be generated.
+		* @param title				A title for the instance to create.
+		* @param rank				The rank index of this interpretation within a stratigraphic column
+		* @param orderingCriteria	How the included horizons are ordered.
+		*/
+		RockFluidOrganizationInterpretation(class OrganizationFeature * orgFeat, const std::string & guid, const std::string & title, ULONG64 index);
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: The index should not be part of the parameters. It must be arbitrary equal to the location of the unit in the organization.
By the way there is a known bug in 2.0.1 about adding fluid units in fuild organization : the workaround is not easy at all to implement!!!!! http://docs.energistics.org/#RESQML/RESQML_TOPICS/RESQML-500-106-0-R-sv2010.html

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#include "resqml2_0_1/AbstractOrganizationInterpretation.h"
+#include "resqml2_0_1/EarthModelInterpretation.h"
+#include "resqml2_0_1/OrganizationFeature.h"
+
+namespace RESQML2_0_1_NS
+{
+	/**
+	* This class is a container for other organizations that are consistent to each others.
+	*/
+	class DLL_IMPORT_OR_EXPORT RockFluidOrganizationInterpretation : public AbstractOrganizationInterpretation
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: You probably miss a pushBackRockFluidUnit in this class to put units in your organization. It can be definitely part of another PR.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		* Check if a grid representation is wether associated to this strati org interp or not.
+		* @param gridRep	The grid representation to check its assocaition with this strati org interp.
+		* @return			True or false.
+		*/
+		bool isAssociatedToGridRepresentation(RESQML2_NS::AbstractGridRepresentation* gridRep) const;
+
+		static const char* XML_TAG;
+		virtual std::string getXmlTag() const {return XML_TAG;}
+
+		ULONG64 getRockFluidUnitInterpretationIndex() const;
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: Should be deleted. It has no meaning. It is a bug in the datamodel : http://docs.energistics.org/#RESQML/RESQML_TOPICS/RESQML-500-106-0-R-sv2010.html

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+{
+	return find(gridRepresentationSet.begin(), gridRepresentationSet.end(), gridRep) != gridRepresentationSet.end();
+}
+
+vector<Relationship> RockFluidOrganizationInterpretation::getAllEpcRelationships() const
+{
+	vector<Relationship> result = AbstractOrganizationInterpretation::getAllEpcRelationships();
+
+	for (size_t i = 0; i < gridRepresentationSet.size(); ++i) {
+		Relationship relRep(gridRepresentationSet[i]->getPartNameInEpcDocument(), "", gridRepresentationSet[i]->getUuid());
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: A more robust way (using misc::getPartNameFromReference(dor)) to achieve the same exists now and avoid to get the Fesapi object.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		relRep.setSourceObjectType();
+		result.push_back(relRep);
+	}
+
+	for (size_t i = 0; i < earthModelSet.size(); i++)
+	{
+		Relationship rel(earthModelSet[i]->getPartNameInEpcDocument(), "", earthModelSet[i]->getUuid());
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: A more robust way (using misc::getPartNameFromReference(dor)) to achieve the same exists now and avoid to get the Fesapi object.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+{
+	vector<Relationship> result = AbstractOrganizationInterpretation::getAllEpcRelationships();
+
+	for (size_t i = 0; i < gridRepresentationSet.size(); ++i) {
+		Relationship relRep(gridRepresentationSet[i]->getPartNameInEpcDocument(), "", gridRepresentationSet[i]->getUuid());
+		relRep.setSourceObjectType();
+		result.push_back(relRep);
+	}
+
+	for (size_t i = 0; i < earthModelSet.size(); i++)
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: ++i is preferred to i++

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+using namespace epc;
+using namespace RESQML2_0_1_NS;
+using namespace gsoap_resqml2_0_1;
+
+const char* RockFluidOrganizationInterpretation::XML_TAG = "RockFluidOrganizationInterpretation";
+
+RockFluidOrganizationInterpretation::RockFluidOrganizationInterpretation(class OrganizationFeature * orgFeat, const std::string & guid, const std::string & title,
+																		 ULONG64 index)
+{
+	if (!orgFeat)
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: check with nullptr  is preferred.

File: src/resqml2_0_1/EarthModelInterpretation.h [hunk 8]
```diff
@@ -73,6 +73,14 @@ namespace RESQML2_0_1_NS
 
 		void pushBackStratiOccurence(class StratigraphicOccurrenceInterpretation * stratiOccurence);
 
+		// Rock Fluid
+		void setRockFluidOrganizationInterpretation(class RockFluidOrganizationInterpretation* rockFluid);
+
+		bool hasRockFluid() const;
```
Comment (philippeVerney) [inline src/resqml2_0_1/EarthModelInterpretation.h]: Naming should be consistent.
hasRockFluid -> hasRockFluidOrganizationInterpretation
getRockFluid -> getRockFluidOrganizationInterpretation

File: src/resqml2_0_1/EarthModelInterpretation.h [hunk 9]
```diff
@@ -52,10 +52,10 @@ namespace RESQML2_0_1_NS
 		*/
 		~EarthModelInterpretation() {}
 
+		// Strati column
```
Comment (philippeVerney) [inline src/resqml2_0_1/EarthModelInterpretation.h]: To delete or to move

File: src/resqml2/AbstractGridRepresentation.h [hunk 10]
```diff
@@ -361,6 +361,51 @@ namespace RESQML2_NS
...
+		* @param stratiUnitIndices	Index of the stratigraphic unit of a given stratigraphic column for each cell. Array length is the number of cells in the grid or the blocked well.
+		* @param nullValue			The value which is used to tell the association between a cell and strati unit is unavailable.
+		* @param stratiOrgInterp	The stratigraphic organization interpretation which is associated to this grid representation.
+		*/
+		void setCellAssociationWithFluidOrganizationInterpretation(ULONG64 * fluidUnitIndices, const ULONG64 & nullValue, RESQML2_0_1_NS::RockFluidOrganizationInterpretation* stratiOrgInterp);
+
+		/**
+		* @return	nullptr if no stratigraphic organization interpretation is associated to this grid representation. Otherwise return the associated stratigraphic organization interpretation.
+		*/
+		RESQML2_0_1_NS::RockFluidOrganizationInterpretation* getFluidOrganizationInterpretation() const;
```
Comment (philippeVerney) [inline src/resqml2/AbstractGridRepresentation.h]: I think I would have preferred to keep "Rock" before "fluid". Same everywhere.
If you want to shorten the method: organization could be "org", interpretation could be "interp"

**Comment by philippeVerney** [comment]
> For now I have no loaded grids into RockFluidOrganizationInterpretation objects and I do not know why

I have to test : I have not seen the bug by simply reading.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp (hunk 4 above)
Comment (untereiner) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: how can I get the Dor of a AbstractObject* ?

Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: Oups my mistake, sorry. You cannot get the Dor of an AbstractObject.
You cannot use misc::getPartNameFromReference for backward relationships. You can only do that for forward relationships.
I'll have to look again at the fesapi mechanism. That's on my plate, definitely not yours.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h (hunk 1 above)
Comment (untereiner) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: Maybe we first make the simple case to work :)

File: src/resqml2/AbstractGridRepresentation.cpp [hunk 11]
```diff
@@ -74,15 +76,14 @@ vector<Relationship> AbstractGridRepresentation::getAllEpcRelationships() const
...
 
 	//Fluid
-	if (hasCellFluidPhaseUnitIndices()) {
-		std::cout << "youpieeeeee" << std::endl;
-		RESQML2_0_1_NS::RockFluidOrganizationInterpretation* fluidOrg = getFluidOrganizationInterpretation();
-		Relationship relFluid(fluidOrg->getPartNameInEpcDocument(), "", fluidOrg->getUuid());
-		relFluid.setDestinationObjectType();
-		result.push_back(relFluid);
+	if(hasCellFluidPhaseUnitIndices()) {
+		if (interp->CellFluidPhaseUnits->FluidOrganization != nullptr) {
```
Comment (philippeVerney) [inline src/resqml2/AbstractGridRepresentation.cpp]: This check is not needed. The XML schema (GSOAP implementation) validates that before FEASPI objects creation.
Looking at annotation of  gsoap proxy interp->CellFluidPhaseUnits->FluidOrganization attribute tells you it is mandatory (and XML schema and UML)

File: src/resqml2/AbstractGridRepresentation.h [hunk 12]
```diff
@@ -361,6 +361,51 @@ namespace RESQML2_NS
 		*/
 		ULONG64 getCellStratigraphicUnitIndices(ULONG64 * stratiUnitIndices);
 
+		//************************************************************
```
[the remark this answers is not in the document]
Comment (philippeVerney) [reply src/resqml2/AbstractGridRepresentation.h]: Actually, my sentence was not a rigorous one but a hint.
You should now replace "rockFluidgraphic" by "Rock fluid" and it should be OK most of time.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h (hunk 2 above)
Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: I agree with "Maybe we first make the simple case work :)".
So let's not think you can add units in your organization as a first step (which is strange from a business point of view)

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h (hunk 3 above)
Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: I confirm it must be deleted.
It is easier to delete and it is sure that we would not need it longer term (For example RESQML2.2 has not even no more this attribute).

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp (hunk 5 above)
Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: My mistake

File: src/common/EpcDocument.h [hunk 13]
```diff
@@ -100,7 +100,6 @@ namespace RESQML2_0_1_NS
 	class WellboreInterpretation;
 	class EarthModelInterpretation;
 	class StructuralOrganizationInterpretation;
-	class RockFluidOrganizationInterpretation;
```
Comment (philippeVerney) [inline src/common/EpcDocument.h]: I am surprised it is not needed. But if not, I am ok.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h (hunk 1 above)
Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: I agree but please : 

- remove the index from the method signature because it will NEVER be useful
- set it arbitrarily to 0 for now on the implementation side.

File: src/resqml2/AbstractGridRepresentation.cpp (hunk 11 above)
Comment (untereiner) [reply src/resqml2/AbstractGridRepresentation.cpp]: I do not understand this part. I have to use interp to do `			Relationship rel(misc::getPartNameFromReference(interp->CellFluidPhaseUnits->FluidOrganization), "", interp->CellFluidPhaseUnits->FluidOrganization->UUID);
`
But in the previous review you told me that in the partial case this will seg fault.

File: src/common/EpcDocument.h (hunk 13 above)
Comment (untereiner) [reply src/common/EpcDocument.h]: It was a duplicate one (see line 128)

File: src/resqml2/AbstractGridRepresentation.cpp (hunk 11 above)
Comment (untereiner) [reply src/resqml2/AbstractGridRepresentation.cpp]: For this part is the code below correct:
`	if(hasCellFluidPhaseUnitIndices()) {
		Relationship rel(misc::getPartNameFromReference(interp->CellFluidPhaseUnits->FluidOrganization), "", interp->CellFluidPhaseUnits->FluidOrganization->UUID);
		rel.setDestinationObjectType();
		result.push_back(rel);
	}`

Comment (philippeVerney) [reply src/resqml2/AbstractGridRepresentation.cpp]: The check is not needed (the if). That's why I thought "interp" could go away.
But, "interp" is actually needed for the line below. It should be named "rep" or "gridRep", my mistake (again).

File: src/common/EpcDocument.h (hunk 13 above)
Comment (philippeVerney) [reply src/common/EpcDocument.h]: Thanks

File: src/resqml2/AbstractGridRepresentation.cpp (hunk 11 above)
Comment (philippeVerney) [reply src/resqml2/AbstractGridRepresentation.cpp]: Yes the proposed code looks good to me.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h (hunk 3 above)
Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: still unresolved

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h (hunk 1 above)
Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: still unresolved

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h (hunk 3 above)
Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: Actually it was resolved in commit 57b953a

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h (hunk 1 above)
Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: Actually it was resolved in commit 57b953a

**Comment by untereiner** [comment]
I'd like to keep "CellFluidPhaseUnitIndices" (without Rock) because the attribute's name is CellFluidPhaseUnitIndices

**Comment by philippeVerney** [comment]
> 
> 
> I'd like to keep "CellFluidPhaseUnitIndices" (without Rock) because the attribute's name is CellFluidPhaseUnitIndices

OK with me

**Comment by philippeVerney** [comment]
The example failed on all platforms but the CI ckecks did not complain because example.exe catched the exception and returned normally.
I have worked on it. It looks it misses all the (or the single in a first step) rock fluid unit management within the rock fluid organization interpretation. A single and unique rock fluid unit is mandatory per RESQML convention. The current exported file is invalid.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp [hunk 14]
```diff
@@ -28,28 +28,34 @@ using namespace gsoap_resqml2_0_1;
...
-
 unsigned int RockFluidOrganizationInterpretation::getGridRepresentationCount() const
 {
-	return gridRepresentationSet.size();
+	const size_t count = gridRepresentationSet.size();
+	if (count > (std::numeric_limits<unsigned int>::max)()) {
+		throw range_error("Too much grid representations.");
+	}
+
+	return static_cast<unsigned int>(gridRepresentationSet.size());
```
Comment (untereiner) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: you could return count ?

**Comment by untereiner** [comment]
something is still wrong in the implementation ?

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp (hunk 14 above)
Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: Technically yes.

But I think I can have warnings from some compiler which did not detect the check above and warn me there could be some potential range error.
By casting, I think I remove these potential warnings.

**Comment by philippeVerney** [comment]
Yes the units of the organization are not at all exported. And you need AT LEAST one per RESQML datamodel rule in a rock fluid organization interp. (you can look at your final build console in travis and appveyor and soon in mines : there is also another tiny error coming before)

If you try to validate the EPC document which results from the example executable, you will see that it is not a valid export.
The reason is because your PR for now focus on link between grid and organization but do not take into account the units which are mandatory.

I am about to tell you that going without the workaround http://docs.energistics.org/#RESQML/RESQML_TOPICS/RESQML-500-106-0-R-sv2010.html as a first step will take more time longer term but I am not totally sure. We can continue to do the first step.
Anyhow, this is really a buggy part of the RESQML datamodel, not at all the best one to start to code in fesapi.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	if (!orgFeat->isPartial() && orgFeat->getKind() != gsoap_resqml2_0_1::resqml2__OrganizationKind__fluid) {
+		throw invalid_argument("The kind of the organization feature is not a fluid organization.");
+	}
+
+	gsoapProxy2_0_1 = soap_new_resqml2__obj_USCORERockFluidOrganizationInterpretation(orgFeat->getGsoapContext(), 1);
+	_resqml2__RockFluidOrganizationInterpretation* rfoi = static_cast<_resqml2__RockFluidOrganizationInterpretation*>(gsoapProxy2_0_1);
+	rfoi->RockFluidUnitIndex = soap_new_resqml2__RockFluidUnitInterpretationIndex(orgFeat->getGsoapContext(), 1);
+	// No need to initialize index since it is a bug : http://docs.energistics.org/#RESQML/RESQML_TOPICS/RESQML-500-106-0-R-sv2010.html
+
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: Probably here, a link to a geologicUnitInterpretation is needed. It should even be a its RockFluidUnitInterpretation specialization.

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+#include "tools/Misc.h"
+
+using namespace std;
+using namespace epc;
+using namespace RESQML2_0_1_NS;
+using namespace gsoap_resqml2_0_1;
+
+const char* RockFluidOrganizationInterpretation::XML_TAG = "RockFluidOrganizationInterpretation";
+
+RockFluidOrganizationInterpretation::RockFluidOrganizationInterpretation(OrganizationFeature * orgFeat, const std::string & guid, const std::string & title)
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: Probably to add a RockFluidUnitInterpretation as another mandatory parameter in this constructor.

Notice that is is to try to go "quickly"!!!!! This is not the long term implementation (several implementations) which should implement the workaround http://docs.energistics.org/#RESQML/RESQML_TOPICS/RESQML-500-106-0-R-sv2010.html (i.e. not really using RockFluidOrganizationInterperation at all but StratigraphicOccurrenceInterpretation instead)

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp (hunk 14 above)
Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: oups sorry.... I understood you wanted to remove the cast.
Actually you wanted to remove the size method. And you are perfectly right. Thanks

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp (hunk 15 above)
Comment (untereiner) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: the class geologicUnitInterpretation does not exist. Should I create it ?

Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidOrganizationInterpretation.cpp]: Up to you.
That's better long term but it is not needed for now (it can definitely be part of another PR). It is quicker not to create GeologicUnitInterpretation.
Having/Creating a RockFluidUnitInterpretation is the most important in my opinion. It can derive directly from AbstractFeatureInterpretation as StratigraphicUnitInterpretation currently does.

**Comment by untereiner** [comment]
I am stuck with two constraints violations. I do not understand how to continue. I will wait until April.

**Comment by philippeVerney** [comment]
Thanks for the work done Lionel. I'll look at it when I'll have time.
By the way, I cannot guarantee to work on it in April. I just know that it is really hard in March. If you want some guarantees, only joining the Fesapi initiative can give you that.

**Comment by philippeVerney** [comment]
I am quickly looking at it and the RESQML datamodel is really a mess.....
A rock fluid unit feature MUST have a top and a bottom boundary feature.....
This is resolved (i.e removed) in RESQML 2.2 but for 2.0.1, it must be satisfied...
I think this is the worst part of the RESQML 2.0.1 datamodel to support.

File: src/resqml2_0_1/RockFluidUnitFeature.cpp [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+-----------------------------------------------------------------------*/
+#include "resqml2_0_1/RockFluidUnitFeature.h"
+
+using namespace std;
+using namespace RESQML2_0_1_NS;
+using namespace gsoap_resqml2_0_1;
+
+const char* RockFluidUnitFeature::XML_TAG = "RockFluidUnitFeature";
+
+RockFluidUnitFeature::RockFluidUnitFeature(soap* soapContext, const string & guid, const string & title)
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidUnitFeature.cpp]: Must add a FluidBoundaryTop and a FluidBoundaryBottom as parameters of this constructor which should be of type "FuildBoundaryFeature"

Comment (untereiner) [reply src/resqml2_0_1/RockFluidUnitFeature.cpp]: thank you! I will had these

Comment (philippeVerney) [reply src/resqml2_0_1/RockFluidUnitFeature.cpp]: The phase attribute is also mandatory and should be added to the constructor parameter list

Comment (untereiner) [reply src/resqml2_0_1/RockFluidUnitFeature.cpp]: ok!

**Comment by untereiner** [comment]
Thank you Philippe! The simplified version of the RockFluid should work (compilation & example)

**Comment by philippeVerney** [comment]
Build and xml part are now ok I think.
But EPC relationships part is not. And fesapi does not validate that (because fesapi does not use that) and it is part of the standards.

getAllEpcRelationhips and importRelationshipSetFromEpc have not been correctly updated for :
- FluidBoundaryFeature (now is linked to RockFluidUnitFeature) and viceversa
- eartModelInterp (now is linked to RockFluidORgInterp) and viceversa

You would face those problems if you would have a more complete deserializeRockFluidOrganization where you try to read all that you have written.
I think/hope this is the last step

**Comment by untereiner** [comment]
Ha, I was happy a bit too fast.

File: src/resqml2_0_1/FluidBoundaryFeature.cpp [hunk 18]
```diff
@@ -17,16 +17,18 @@ specific language governing permissions and limitations
...
 
 using namespace std;
 using namespace RESQML2_0_1_NS;
 using namespace gsoap_resqml2_0_1;
+using namespace epc;
 
 const char* FluidBoundaryFeature::XML_TAG = "FluidBoundaryFeature";
 
-FluidBoundaryFeature::FluidBoundaryFeature(soap* soapContext, const string & guid, const string & title, const resqml2__FluidContact & fluidContact)
+FluidBoundaryFeature::FluidBoundaryFeature(soap* soapContext, const string & guid, const string & title, const resqml2__FluidContact & fluidContact) : rockFluidUnitFeature(nullptr)
```
Comment (philippeVerney) [inline src/resqml2_0_1/FluidBoundaryFeature.cpp]: rockFluidUnitFeature should be a vector (there can be several units using this boundary as a top or bottom).
Ideally, this vector should be in BoundaryFeature superclass according to RESQML datamodel

File: src/resqml2_0_1/FluidBoundaryFeature.cpp [hunk 19]
```diff
@@ -39,4 +41,14 @@ FluidBoundaryFeature::FluidBoundaryFeature(soap* soapContext, const string & gui
 	setMetadata(guid, title, std::string(), -1, std::string(), std::string(), -1, std::string());
 }
 
+std::vector<epc::Relationship> FluidBoundaryFeature::getAllEpcRelationships() const
+{
+	vector<Relationship> result = BoundaryFeature::getAllEpcRelationships();
+
```
Comment (philippeVerney) [inline src/resqml2_0_1/FluidBoundaryFeature.cpp]: The source rel with rock fluid unit feature should be listed here.
Ideally it should be done in BoundaryFeature superclass

File: src/resqml2_0_1/FluidBoundaryFeature.h [hunk 20]
```diff
@@ -55,6 +56,16 @@ namespace RESQML2_0_1_NS
...
 		static const char* XML_TAG;
 		virtual std::string getXmlTag() const {return XML_TAG;}
-	};
+
+	private:
+		std::vector<epc::Relationship> getAllEpcRelationships() const;
+		void importRelationshipSetFromEpc(COMMON_NS::EpcDocument* epcDoc);
+
+		// backward relationship
+		RockFluidUnitFeature* rockFluidUnitFeature;
```
Comment (philippeVerney) [inline src/resqml2_0_1/FluidBoundaryFeature.h]: Ideally it should be done in BoundaryFeature superclass

File: src/resqml2_0_1/FluidBoundaryFeature.h [hunk 21]
```diff
@@ -55,6 +56,16 @@ namespace RESQML2_0_1_NS
...
+
+		friend void RockFluidUnitFeature::setFluidBoundaryTop(FluidBoundaryFeature* fluidBoundaryFeature);
```
Comment (philippeVerney) [inline src/resqml2_0_1/FluidBoundaryFeature.h]: Ideally it should be done in BoundaryFeature superclass

File: src/resqml2_0_1/FluidBoundaryFeature.h [hunk 22]
```diff
@@ -55,6 +56,16 @@ namespace RESQML2_0_1_NS
 
 		static const char* XML_TAG;
 		virtual std::string getXmlTag() const {return XML_TAG;}
-	};
+
+	private:
+		std::vector<epc::Relationship> getAllEpcRelationships() const;
```
Comment (philippeVerney) [inline src/resqml2_0_1/FluidBoundaryFeature.h]: Not really needed as it should be done in BoundaryFeature superclass

File: src/resqml2_0_1/FluidBoundaryFeature.h [hunk 23]
```diff
@@ -55,6 +56,16 @@ namespace RESQML2_0_1_NS
...
+		void importRelationshipSetFromEpc(COMMON_NS::EpcDocument* epcDoc);
```
Comment (philippeVerney) [inline src/resqml2_0_1/FluidBoundaryFeature.h]: Not really needed as it should be done in BoundaryFeature superclass

File: src/resqml2_0_1/RockFluidOrganizationInterpretation.h [hunk 24]
```diff
@@ -82,6 +82,9 @@ namespace RESQML2_0_1_NS
 		*/
 		DLL_IMPORT_OR_EXPORT void pushBackRockFluidUnitInterpretation(class RockFluidUnitInterpretation * rockFluidUnitInterpretation);
 
+		DLL_IMPORT_OR_EXPORT bool hasRockFluidUnitInterpretation() const;
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidOrganizationInterpretation.h]: Not sure it is needed since that's the buggy part. It honors the bug.

File: src/resqml2_0_1/RockFluidUnitFeature.cpp [hunk 25]
```diff
@@ -39,3 +41,96 @@ RockFluidUnitFeature::RockFluidUnitFeature(soap* soapContext, const string & gui
...
+	_resqml2__RockFluidUnitFeature* interp = static_cast<_resqml2__RockFluidUnitFeature*>(gsoapProxy2_0_1);
+
+	if (interp->FluidBoundaryTop != nullptr)
+	{
+		Relationship rel(misc::getPartNameFromReference(interp->FluidBoundaryTop), "", interp->FluidBoundaryTop->UUID);
+		rel.setDestinationObjectType();
+		result.push_back(rel);
+	}
+
+	if (interp->FluidBoundaryBottom != nullptr)
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidUnitFeature.cpp]: This if is not needed since it is forced to exist by the RESQML datamodel.

File: src/resqml2_0_1/RockFluidUnitFeature.h [hunk 26]
```diff
@@ -50,7 +50,18 @@ namespace RESQML2_0_1_NS
 		*/
 		~RockFluidUnitFeature() {}
 
+		DLL_IMPORT_OR_EXPORT void setFluidBoundaryTop(class FluidBoundaryFeature* fluidBoundaryFeature);
```
Comment (philippeVerney) [inline src/resqml2_0_1/RockFluidUnitFeature.h]: Ideally, it should not constrain to use FluidBoundaryFeature. It would be better to use its BoundaryFeature super class.
